[Jetpack plugin install prompt] Improve prompt presentation logic and handle install cancellation

staskus opened the pull request:
This is part of [Show Jetpack plugin install prompt during Jetpack app login process](https://github.com/wordpress-mobile/WordPress-iOS/issues/19213). Merging to the parent branch.

## Description

### 1. Making improvements to Jetpack plugin install prompt presentation logic:
- Fixing the issue that installation flow (JetpackLoginViewController) couldn't be dismissed in Stats and Notifications
- Moving Jetpack install flow logic from `JetpackLoginViewController` to `JetpackInstallCoordinator` so it could be reused in `JetpackInstallPromptViewController`
- Updating `JetpackRemoteInstallViewController` so it would display "Back" button until installation begins
- Moving Jetpack Prompt presentation logic to `MySitesViewController`
- Presenting Jetpack plugin install prompt modally and not pushing it into login flow, to indicate that one flow ended and the other one started

![image](https://user-images.githubusercontent.com/4062343/187418802-f4533f60-7888-48df-8cb3-eda429b53d15.png)

### 2. Added feature flag

Adding feature flag so we could merge this branch and it would be only visible for local development and PR branch testing.

## Testing instructions

### Case 0: Jetpack plugin prompt appears
1. Fresh install Jetpack App
2. Log into a self-hosted site that doesn't have the Jetpack plugin installed

### Case 1: Jetpack plugin is installed + WP.com logged in
1. Execute Case 0
2. Click "Install"
3. Follow the whole installation flow
4. Log in to WP.com account when prompted
5. After successful log in, the install prompt is dismissed and the main screen appears

### Case 2: Come back from installation
1. Execute Case 0
2. Click "Install"
3. Click "Back" button
4. You should come back to the install prompt

### Case 3: Cancel installation
1. Execute Case 0
2. Click "Install"
3. Click "Continue"
4. Notice "Cancel" button appears once the installation begins
5. Click "Cancel"
6. Home screen appears

### Case 4: Learn more
1. Execute Case 0
7. Click "Learn More"
8. Jetpack website should open

### Case 5: No thanks
1. Execute Case 0
6. Click "No thanks"
7. Prompt should be dismissed

**Note:** [installation issues noted in Phase 3 can be ignored](https://github.com/wordpress-mobile/WordPress-iOS/issues/17721)

## Regression Notes

1. Potential unintended areas of impact

Other login flows, since WordPressAuthenticationManager and JetpackLoginViewController were edited

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- Run Unit tests
- Run UI tests
- Tested WP.com and social log-in manually

3. What automated tests I added (or what prevented me from doing so)

- UI tests are in the plans of the project.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Videos

### Install prompt flow

https://user-images.githubusercontent.com/4062343/187420228-c50818f3-20de-4e02-ab8d-827593bfd0ca.mov

staskus committed:
Consolidating Jetpack prompt presentation logic

Until now, presentation of JetpackInstallPromptViewController and  JetpackLoginViewController was split between MySiteViewController and WordPressAuthenticationManager, making it hard to control and reason about.

1. Making changes to present the prompt modally (on top), instead of pushing it on top of Login flow. It signifies to user the end of one flow and the start of another.
2. After presenting JetpackLoginViewController (installation flow), we need to handle the completion:
- If installation flow is fully completed - dismiss the prompt
- If installation flow is partially completed (Jetpack connected) - dismiss the prompt
- If installation flow is cancelled - continue showing the prompt, giving user the chance to learn more

staskus committed:
Do not dismiss JetpackInstallPromptViewController by itself

Delegating the dismissal of the prompt to the presenter

staskus committed:
Fixing Jetpack plugin install dismissing in Stats

Reverting the change made previously. If completionBlock existed, the dismiss was not called.

staskus committed:
Removing Jetpack prompt presentation logic from WordPressAuthenticationManager

- Continue posting the notification (as done with previous prompts)
- Posting the notification when login is dismissed

guarani commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController+JetpackInstall.swift:
```diff
@@ -2,34 +2,66 @@ import UIKit
+        }
 
-        navigationController?.present(navController, animated: true)
+        navigationController?.present(installPromptNavigationController, animated: true)
```
```suggestion
        present(installPromptNavigationController, animated: true)
```

Is there an advantage to presenting from the nav controller rather than its root view controller?

staskus replied:
> I notice that the "what would you like to focus on" question is shown after Jetpack setup is complete. This might be out of scope, but I didn't see it on the link to the known "existing issues".

Yes, it appears after WP.com login. As I understood it's mentioned under `The login epilogue can be displayed during the connection process, we should prevent it from displaying`

> Is there an advantage to presenting from the nav controller rather than its root view controller?

There's no advantage or disadvantage. Those 2 calls are equivalent, since MySiteViewController is inside `UINavigationController`, so other controllers are presented in & on `UINavigationController` either way. I'll remove the redundant `navigationController?.` code.

> When I finish the whole setup, both the installation flow and the prompt flow modals are dismissed in quick succession. Should we instead dismiss both modals in one single animation?
Are there too many modals? In case 2, the user dismisses two modals and the app dismisses another (total 3 modals).

Yes, it would look better 2 dismiss them with a single animation.

It's a good question if there're too many modals. Installation flow by itself is designed to be presented in a modal (with all the dismissal cancellation buttons in the navigation bar) and WP.com login flow is presented as a modal as well. 

I'll check what changes I'd need to make if we push the installation flow instead of presenting it on top of the installation prompt.

guarani replied:
> Yes, it appears after WP.com login. As I understood it's mentioned under `The login epilogue can be displayed during the connection process, we should prevent it from displaying`

I'd interpreted the epilogue as being the site picker, but I agree it also covers this focus question screen.

> There's no advantage or disadvantage. Those 2 calls are equivalent, since MySiteViewController is inside `UINavigationController`, so other controllers are presented in & on `UINavigationController` either way. I'll remove the redundant `navigationController?.` code.

Thanks for explaining, I wasn't familiar with calling `present` from a navigation controller.

> I'll check what changes I'd need to make if we push the installation flow instead of presenting it on top of the installation prompt.

Ok 👍

staskus committed:
Added feature flag for Jetpack plugin install prompt presentation

guarani commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController+JetpackInstall.swift:
```diff
@@ -2,34 +2,66 @@ import UIKit
-        let controller = JetpackLoginViewController(blog: blog)
-        controller.promptType = .installPrompt
+        let installPromptViewController = JetpackInstallPromptViewController(blog: blog)
+        let installPromptNavigationController = UINavigationController(rootViewController: installPromptViewController)
```
It looks like this prompt screen is "standalone" because the installation flow is another modal on top of it (it's not pushed onto the nav controller created on the line above). If it's "standalone", is the nav controller needed?

staskus replied:
Good observation 👍. In the current situation, it's currently not needed, since we don't have any navigation flow inside the prompt.

guarani commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController+JetpackInstall.swift:
```diff
@@ -2,34 +2,66 @@ import UIKit
+            case .noThanks:
+                installPromptNavigationController.dismiss(animated: true)
+            case .install:
+                self?.presentJetpackInstall(on: installPromptNavigationController, blog: blog, jetpackInstallPromptSettings: promptSettings)
```
Similar to the question above about the need for a navigation controller, should we push this installation flow screen on to`installPromptNavigationController` instead of presenting it on top of it? 

I could see a reason to keep it this way if we want users who cancel some part of the installation flow to return to the prompt. However, as a user, I think if I canceled the installation flow at any point, I'd also want to dismiss the prompt. Wdyt?

staskus replied:
This is a valid question. 

If we push the installation flow, I thought then it would make sense to allow users to come `< Back` rather than dismiss. Although not allowing them to come back would allow avoiding these additional "partial installation" checks inside the prompt. I'll think about it.

guarani commented on WordPress/Classes/ViewRelated/Blog/My Site/MySiteViewController+JetpackInstall.swift:
```diff
@@ -2,34 +2,66 @@ import UIKit
+            /// If Jetpack install + WP.com login flow is completed, the blog will be synced and jetpack.isConnected will be true
+            /// Dismissing the prompt
+            if let jetpack = blog.jetpack, jetpack.isConnected {
+                self?.syncBlogs()
```
Would this try to sync WP.com blogs even if Jetpack is not connected to a WP.com account? My understanding from reading the code today is that `isConnected` means Jetpack is connected to WP.com, but not connected to a WP.com account.

While I tested the test case where the user cancels the installation flow just before connecting their account, I haven't yet had time to test if this code is executed or if it's a different flow.

staskus replied:
Yes. We do this syncing to know if the blog actually became `isConnected` or not, after canceling the flow.  

> My understanding from reading the code today is that isConnected means Jetpack is connected to WP.com, but not connected to a WP.com account.

Yes, this is what it means.

guarani reviewed (commented):
Looking good! 

I added some questions above to better understand some parts of the code. I also have a couple of more generic questions:
- I notice that the "what would you like to focus on" question is shown after Jetpack setup is complete. This might be out of scope, but I didn't see it on the link to the known "existing issues".
- When I finish the whole setup, both the installation flow and the prompt flow modals are dismissed in quick succession. Should we instead dismiss both modals in one single animation?
- Are there too many modals? In case 2, the user dismisses two modals and the app dismisses another (total 3 modals).

staskus committed:
Remove redundant navigationController call when presenting

staskus commented:
**Summary**

I agree that we should not present so many levels of modals and it's better to integrate installation flow inside the installation prompt.

It will require refactoring of `JetpackLoginViewController`. I was reluctant to do so but the way it's made makes it very difficult to reuse or control. Right now this UIViewController is:
- Checking for `installPrompt` flag inside `JetpackLoginViewController` to immediately present `JetpackRemoteInstallViewController` in `viewDidLoad`. It's hard to understand when looking and the code, and doesn't allow to control the mode of presentation from the installation flow.
- Coordinating installation flow by acting as a delegate of multiple installation view controllers and steps.
- Shows UI to "Install" Jetpack. Essentially acting as an install prompt for Stats and Notifications controllers.

We should:
- Move install flow coordination logic from `JetpackLoginViewController` to a common place
- Leave `JetpackLoginViewController` only responsible for showing UI to "Install" Jetpack and using common coordination logic
- Use common coordination logic in our login install prompt

staskus committed:
Moving installation flow logic to JetpackInstallCoordinator

- Moving installation flow and presentation logic from JetpackLoginViewController to JetpackInstallCoordinator
- JetpackInstallCoordinator presents Jetpack plugin installation and Jetpack connection controllers, does the tracking and calls completion blocks
- Using JetpackInstallCoordinator in JetpackLoginViewController

staskus committed:
Show cancel button when installation begins

When installation controller is presented and it's not the only view controller, keep showing back button until installation begins

staskus committed:
Using JetpackInstallCoordinator in JetpackInstallPromptViewController

staskus committed:
Simplify presentation of JetpackInstallPrompt

The install coordination is now managed by JetpackInstallCoordinator and coming back to the prompt is not allowed, so we don't need to handle corner cases anymore

staskus commented:
Updated description (test cases, videos) based on our discussion and refactoring

staskus commented:
@sla8c would you be able to review this PR? I'm interested in the feedback.

staskus committed:
Merge branch 'feature/19213-jetpack-install-prompt' into feature/jetpack-plugin-install-prompt-integration-into-current-flow

staskus committed:
Merge branch 'feature/19213-jetpack-install-prompt' into feature/jetpack-plugin-install-prompt-integration-into-current-flow

guarani commented:
👋 Hi @staskus, I haven't had a chance to review this today. I'll review it ASAP tomorrow.

staskus commented:
> 👋 Hi @staskus, I haven't had a chance to review this today. I'll review it ASAP tomorrow.

Hi, @guarani 👋, all good, it might not be necessary to review it ASAP. I'll go through it with @sla8c today.

staskus committed:
Make JetpackInstallPromptViewController immutable

Use delegate instead of dismiss closure to keep JetpackInstallPromptViewController immutable

sla8c commented:
> > 👋 Hi @staskus, I haven't had a chance to review this today. I'll review it ASAP tomorrow.
> 
> Hi, @guarani 👋, all good, it might not be necessary to review it ASAP. I'll go through it with @sla8c today.

Hi @Guarani - @staskus and I had been discussing and I had seen the opportunity for a pairing session + enhancement for this PR which we did today (just now) thx

guarani commented:
Cool! Thanks for letting me know @sla8c 👍 
I understand this to mean you'll re-request a review when needed @staskus. If we're blocked by design though, I'm happy to discuss and see what options are available.

staskus commented:
@sla8c 

If you think that the code changes themselves are okay, we could merge them to the parent branch. When the UX feedback for the prompt comes, I'll make another PR.

sla8c commented:
@staskus I'm fine to approve but was waiting for all the tests to pass (I just checked now and they are still not passing).  Can you please look at them + resolve them and then I'll approve? thanks 🙂

staskus committed:
Merge branch 'trunk' into feature/jetpack-plugin-install-prompt-integration-into-current-flow

staskus committed:
Merge branch 'feature/19213-jetpack-install-prompt' into feature/jetpack-plugin-install-prompt-integration-into-current-flow

staskus commented:
💡 @wordpress-mobile/mobile-ui-testing-squad No need for a review. It was requested by accident.

staskus commented:
@sla8c the tests passed. Merging to the latest trunk was enough.
